### ISSUE TITLE
Unexpected errors with reportImplicitOverride 

### ISSUE THREAD

**Issue by mehdigmira** [microsoft/pyright#4905]
**Describe the bug**
I've tried to enable `reportImplicitOverride` flag, but it's returning errors on some cases which were unexpected to me:
- Any `__init__` function is `Method "__init__" is not marked as override but is overriding a method in class "object"`. While this is technically true, I find it surprising that there not any special casing for `object`
```python
# pyright: strict
# pyright: reportImplicitOverride=true

class A:
    def __init__(self) -> None:  # Method "__init__" is not marked as override but is overriding a method in class "object"
        ...
```
- Seems there is bug when code involves some Unknown classes, see code below
```python
from __future__ import annotations

from typing import TYPE_CHECKING
from sqlalchemy.orm import RelationshipProperty. # I've used this as an example, but i guess this would work with anything else


# pyright: strict
# pyright: reportImplicitOverride=true


if TYPE_CHECKING:
    X = int
else:

    class X(RelationshipProperty):  # RelationshipProperty: Type[RelationshipProperty[Unknown]]
        ...


def f():  # Method "f" is not marked as override but is overriding a method in class "Unknown"
    ...
```

**To Reproduce**
For the second example you may want to install https://github.com/sqlalchemy/sqlalchemy2-stubs

**Comment by ifuatgucluer** [comment]
good job!

**Comment by erictraut** [comment]
On the first point, this was changed in pyright 1.1.302, which was published yesterday. For details, refer to [this issue](https://github.com/microsoft/pyright/issues/4823). Support for the `reportImplicitOverride` is based on the new [PEP 698](https://peps.python.org/pep-0698/), and there were parts of it that were unclear — including the point about whether `__init__` and `__new__` should be exempted from strict override checks. The author of the PEP clarified his intention, and I updated pyright accordingly.

As for your second point, I don't understand the example code. The function `f` isn't a method, so it is not affected by the `reportImplicitOverride` check. 

Perhaps you meant to show something like this, where a class derives from some unknown class (e.g. a symbol that is imported from a library with no type information)?

```python
# pyright: reportImplicitOverride=true

from foo import Foo


class Bar(Foo):
    def f(self) -> None:
        ...
```

In this case, `Foo` has the type `Unknown`, so a type checker has no way to determine whether the method `f` is overriding a method of the same name in its base class. The `reportImplicitOverride` check therefore shouldn't report anything. This is as designed.

I think pyright is working as intended here, so I'm going to close the issue. If there's something I've missed, feel free to respond with additional details.

**Comment by mehdigmira** [comment]
- great for point 1 !
- No the example i gave is complete. The error is reported by pyright, though the function is not within a class, see screenshot below
<img width="651" alt="image" src="https://user-images.githubusercontent.com/13655536/230168491-cfd5a129-c94c-4079-8028-fe30a069b475.png">

**Comment by erictraut** [comment]
Oh, it looks like you're using pylance, which is a few weeks behind pyright in terms of bug fixes. (For future bug reports, it would be useful if you'd include the version of pyright or pylance that you're using.) You'll notice in the screen shot that it's reporting the error using the `reportGeneralTypeIssues` rule rather than `reportImplicitOverride` rule. This was a bug in my initial implementation of this check, and it was fixed a couple of weeks ago in pyright. It should be fixed in this week's release of pylance (which should be published within the next 12 hours).

**Comment by mehdigmira** [comment]
The screenshot was just to show you the error. But i have the same thing using pyright CLI. I was using pyright 1.1.301, since 1.1.302 was not released yet.

But now I'm able to reproduce with pyright 1.1.302 as well:

_file typing_debug.py_
```
from __future__ import annotations

from typing import TYPE_CHECKING

from sqlalchemy.orm import RelationshipProperty

# pyright: strict
# pyright: reportImplicitOverride=true


if TYPE_CHECKING:
    X = int
else:

    class X(RelationshipProperty):  # RelationshipProperty: Type[RelationshipProperty[Unknown]]
        ...


def f():  # Method "f" is not marked as override but is overriding a method in class "Unknown"
    ...
```

_pyright version_
```
$ pyright --version
pyright 1.1.302
```

_running pyright_
```
$ pyright typing_debug.py 
No configuration file found.
pyproject.toml file found at /code.
Loading pyproject.toml file at /code/pyproject.toml
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.302
typing_debug.py
  typing_debug.py:19:5 - error: Method "f" is not marked as override but is overriding a method in class "Unknown" (reportImplicitOverride)
1 error, 0 warnings, 0 informations 
Completed in 1.245sec
```

I'm sorry if i'm missing smth out.

**Comment by mehdigmira** [comment]
The code makes no sense, but it's the minimal code i can provide to reproduce this (at least reproduce on my side). 
- if i remove the if type_checking block all together, no error i reported
- if i keep it but replace X definition by smth not related to RelationshipProperty, no error is reported

I'll try to see if i can a better example, and try to reproduce it in a new python environment with the the most minimal setup

**Comment by mehdigmira** [comment]
Are you able to reproduce with this ? https://github.com/mehdigmira/implicit_override_test_case

**Comment by erictraut** [comment]
Yes, I'm able to repro the problem now. Thanks for taking the time to create a repro case.

The problem turned out to have nothing to do with the error that was being reported, which is why it was so mystifying. Instead, it was a bug that affected any `class` or `def` statement located within a code block that pyright's binder determined to be unreachable. In this case, no "scope" data structure was created for the `class` or `def` statement. Later, when the type evaluator logic was running, it assumed there would be a "scope" (and associated symbol table) for these nodes. Because there was no symbol table present for the class `X`, it ended up using the symbol table for the module(!) instead. This led it to conclude that there was a class `X` that had a method called `f`. In other words, it was very confused at that point, and the internal data structures were corrupted. This is a very old bug. I'm surprised it hasn't been reported in some form previously.

This will be fixed in the next release.

**Comment by iacobfred** [comment]
I still see Pylance reporting errors like `Method "my_method" is not marked as override but is overriding a method in class "Unknown"` if a class inherits from a symbol from a library with no type stubs. Does this mean @erictraut 's change is not yet included in Pylance?

**Comment by erictraut** [comment]
@iacobfred, this diagnostic rule isn't applicable if your class derives from `Unknown`. A If your class derives from `Unknown`, a static type checker has no visibility into the methods of the parent class, so it cannot determine whether a child class is overriding a method. The `reportImplicitOverride` check requires that you have complete and accurate type information for the parent classes in your class hierarchy.

**Comment by iacobfred** [comment]
@erictraut , can you explain what you mean by "if your class derives from `Unknown`"?

My class derives from the `TypedModel` class defined here: [https://github.com/iacobfred/django-typed-models/blob/master/typedmodels/models.py](https://github.com/iacobfred/django-typed-models/blob/master/typedmodels/models.py)

**Comment by erictraut** [comment]
> can you explain what you mean by "if your class derives from `Unknown`"?

That means one or more of the base classes for a class cannot be determined statically for one reason or another. Sometimes this is because an import cannot be resolved, the symbol is missing, it is defined in a way that is too dynamic for a static type checker to infer, etc. If you'd like me to look at your specific example, please open a new discussion topic and provide a minimal, self-contained code sample and repro steps.
